### PR TITLE
Optionally avoid evicting critical pods in kubelet

### PR DESCRIPTION
**Pull request by vishh**
For #40573

```release-note
When feature gate "ExperimentalCriticalPodAnnotation" is set, Kubelet will avoid evicting pods in "kube-system" namespace that contains a special annotation - `scheduler.alpha.kubernetes.io/critical-pod`
This feature should be used in conjunction with the rescheduler to guarantee availability for critical system pods - https://kubernetes.io/docs/admin/rescheduler/
```

### CODE REVIEW

**Comment by derekwaynecarr** [comment]
I would like to review this PR prior to its merging.

File: pkg/features/kube_features.go [hunk 1]
```diff
@@ -58,6 +58,13 @@ const (
 	// contains a privileged container, or specific non-namespaced capabilities (MKNOD, SYS_MODULE,
 	// SYS_TIME). This should only be enabled if user namespace remapping is enabled in the docker daemon.
 	ExperimentalHostUserNamespaceDefaultingGate utilfeature.Feature = "ExperimentalHostUserNamespaceDefaulting"
+
+	// owner: @vishh
+	// alpha: v1.5
+	//
+	// Ensures guaranteed scheduling of pods marked with a special pod annotation `scheduler.alpha.kubernetes.io/critical-pod`
```
Comment (derekwaynecarr) [inline pkg/features/kube_features.go]: i would also clarify here that its restricted to the kube-system namespace.

**Review by derekwaynecarr — CHANGES_REQUESTED** [review]
one question, and a nit

File: pkg/kubelet/eviction/eviction_manager.go [hunk 2]
```diff
@@ -110,7 +111,7 @@ func (m *managerImpl) Admit(attrs *lifecycle.PodAdmitAttributes) lifecycle.PodAd
 	// the node has memory pressure, admit if not best-effort
 	if hasNodeCondition(m.nodeConditions, v1.NodeMemoryPressure) {
 		notBestEffort := v1.PodQOSBestEffort != qos.GetPodQOS(attrs.Pod)
-		if notBestEffort || kubetypes.IsCriticalPod(attrs.Pod) {
+		if notBestEffort {
```
Comment (vishh) [inline pkg/kubelet/eviction/eviction_manager.go]: Yeah. I don't think it makes sense to run "Critical" BestEffort pods

File: pkg/features/kube_features.go (hunk 1 above)
Comment (vishh) [inline pkg/features/kube_features.go]: Ack. I will add the additional comments.

Comment (vishh) [inline pkg/features/kube_features.go]: Done

**Review by derekwaynecarr — APPROVED** [review]

**Comment by vishh** [comment]
@dchen1107 has some concerns with this PR being cherrypicked. She is yet to post her concerns. For that reason, I have removed the cherrypick request label for now.
This PR is ready to go otherwise.

**Comment by dchen1107** [comment]
Like what we discussed offline, I am ok to have this to HEAD, and it is what we discussed offline. @dashpole is going to have some simple preemption logic in Kubelet for 1.6 to cover the rest. Letting Kubelet do some preemption is the direction I tried to push forward, and luckily we reached the agreement now. 

But I am concerned to cherrypick it to either 1.4 or 1.5 branch unless we made some other changes too:

* This one avoid evicting critical pods, no matter which is static or not. This is good. But fluentd is not marked as critical pod yet and running as static pod too. In this case, it might be evicted, and Kubelet never admit it again. You might have to mark fluentd as critical for now to workaround this issue.

* You revert the commit of kube-proxy oom score change. kube-proxy is running as burstable. Without proper preemption logic, it might become the victim by system OOM on the overloaded node all the time. One can argue that before we have the system OOM, the eviction manager can evict some other non-critical pods to prevent such situations. But there are two caveats:

  * We know we disabled on-demand memory notification for 1.5 by default. Without that feature, system OOM will be triggered a lot before the eviction manager can take any action. 
  * Scheduler might pack critical pods to the same node. With your changes, the eviction manager is effectively disabled, thus, kube-proxy might become the victim here. 

I suggested you only cherrypick the feature-gate flag pieces to 1.4 and 1.5. Actually that is what we agreed upon if I remembered it correctly. Or you can either

* Mark fluentd as critical one or
* cherrypick to-be-done preemption handling to 1.4 and 1.5. But I think that would be too risk here. 

cc/ @bprashanth @thockin @kubernetes/sig-network-bugs

**Comment by vishh** [comment]
@dchen1107 

> Mark fluentd as critical one 

Thanks for this tip. I will update fluentd too.

> You revert the commit of kube-proxy oom score change. kube-proxy is running as burstable. Without proper preemption logic, it might become the victim by system OOM on the overloaded node all the time. One can argue that before we have the system OOM

I don't think system OOMs should be an issue because kube-proxy container will get restarted. Only evictions are terminal.

> We know we disabled on-demand memory notification for 1.5 by default. Without that feature, system OOM will be triggered a lot before the eviction manager can take any action.

This is true. But, I'm hoping that kube-proxy does not consume too much above its request. If it does, we should bump its request.

> Scheduler might pack critical pods to the same node. With your changes, the eviction manager is effectively disabled, thus, kube-proxy might become the victim here.

Can you explain a bit more? We might have issues with admission, which we agreed is a `known issue` and we do not intend to fix until we have preemptions. IIUC `Kube-proxy` will never be evicted once admitted and so I don't see any issue.

**Comment by vishh** [comment]
I chatted with @bprashanth and cherry-picking this change without preemptions into v1.5 & has the following issues:
1. It makes dashboard, heapster, kube-dns, etc. unevictable. We are not sure if pods other can kube-proxy can be trusted as of now.
2. System OOMs leading to crash loop backoff - System OOMs cannot be reliably avoided using oom score adjust setting unless we disable system OOMs all together for critical pods. This is dangerous because critical pods can also have memory leak bugs

If we include preemptions, then `1` will no longer be an issue, but `2` will continue to remain an issue. We need to solve `2` in v1.6.
I'm inclined towards not meddling with oom score adjust, which is known to be unreliable, and instead reduce max backoff for crash looping critical pods. 

As for v1.5 and v1.4, the easiest option is to flag gate all the changes that were made to protect kube-proxy using a separate flag. Cherrypicking all the new features is more risky when compared to just adding a flag.

**Pull request merged**